File: serde-generate/runtime/csharp/Serde/BinaryDeserializer.cs
// Copyright (c) Facebook, Inc. and its affiliates
// SPDX-License-Identifier: MIT OR Apache-2.0

using System;
using System.IO;
using System.Numerics;
using System.Text;

namespace Serde
{
    public abstract class BinaryDeserializer : IDeserializer, IDisposable
    {
        protected readonly ArraySegment<byte> input;
        protected readonly BinaryReader reader;
        protected readonly Encoding utf8 = Encoding.GetEncoding("utf-8", new EncoderExceptionFallback(), new DecoderExceptionFallback());
        private long containerDepthBudget;

        public BinaryDeserializer(byte[] _input, long maxContainerDepth) : this(new ArraySegment<byte>(_input), maxContainerDepth) { }

        public BinaryDeserializer(ArraySegment<byte> _input, long maxContainerDepth)
        {
            input = _input;
            reader = new BinaryReader(new MemoryStream(input.Array, input.Offset, input.Count));
            containerDepthBudget = maxContainerDepth;
        }

        public void Dispose() => reader.Dispose();

        public int get_buffer_offset() => (int)reader.BaseStream.Position;

        public abstract long deserialize_len();
        public abstract int deserialize_variant_index();
        public abstract void check_that_key_slices_are_increasing(Range key1, Range key2);

        public char deserialize_char() => throw new DeserializationException("Not implemented: char deserialization");

        public float deserialize_f32() => reader.ReadSingle();

        public double deserialize_f64() => reader.ReadDouble();

        public void increase_container_depth()
        {
            if (containerDepthBudget == 0)
            {
                throw new DeserializationException("Exceeded maximum container depth");
            }
            containerDepthBudget -= 1;
        }

        public void decrease_container_depth()
        {
            containerDepthBudget += 1;
        }

        public string deserialize_str()
        {
            long len = deserialize_len();
            if (len < 0 || len > int.MaxValue)
            {
                throw new DeserializationException("Incorrect length value for C# string");
            }
            byte[] content = reader.ReadBytes((int)len);
            if (content.Length < len)
                throw new DeserializationException($"Need {len - content.Length} more bytes for string");
            return utf8.GetString(content);
        }

        public ValueArray<byte> deserialize_bytes()
        {
            long len = deserialize_len();
            if (len < 0 || len > int.MaxValue)
            {
                throw new DeserializationException("Incorrect length value for C# array");
            }
            byte[] content = reader.ReadBytes((int)len);
            if (content.Length < len)
                throw new DeserializationException($"Need {len - content.Length} more bytes for byte array");
            return new ValueArray<byte>(content);
        }

        public bool deserialize_bool()
        {
            byte value = reader.ReadByte();
            switch (value)
            {
                case 0: return false;
                case 1: return true;
                default: throw new DeserializationException("Incorrect value for bool: " + value);
            }
        }

        public Unit deserialize_unit() => new Unit();

        public byte deserialize_u8() => reader.ReadByte();

        public ushort deserialize_u16() => reader.ReadUInt16();

        public uint deserialize_u32() => reader.ReadUInt32();

        public ulong deserialize_u64() => reader.ReadUInt64();

        public BigInteger deserialize_u128()
        {
            BigInteger signed = deserialize_i128();
            if (signed >= 0)
            {
                return signed;
            }
            else
            {
                return signed + (BigInteger.One << 128);
            }
        }

        public sbyte deserialize_i8() => reader.ReadSByte();

        public short deserialize_i16() => reader.ReadInt16();

        public int deserialize_i32() => reader.ReadInt32();

        public long deserialize_i64() => reader.ReadInt64();

        public BigInteger deserialize_i128()
        {
            byte[] content = reader.ReadBytes(16);
            if (content.Length < 16)
                throw new DeserializationException("Need more bytes to deserialize 128-bit integer");
            return new BigInteger(content);
        }

        public bool deserialize_option_tag()
        {
            byte value = reader.ReadByte();
            switch (value)
            {
                case 0: return false;
                case 1: return true;
                default: throw new DeserializationException("Incorrect value for Option tag: " + value);
            }
        }
    }
}

File: serde-generate/runtime/swift/Serde/BinaryDeserializer.swift
//  Copyright © Diem Association. All rights reserved.

import Foundation

enum BinaryDeserializerError: Error {
  case deserializationException(issue: String)
}

public class BinaryDeserializer : Deserializer {
  
  let reader: BinaryReader
  fileprivate let input: [UInt8]
  fileprivate var containerDepthBudget: Int64
  
  init(input: [UInt8], maxContainerDepth: Int64) {
    self.input = input
    self.reader = BinaryReader(data: Data(input) as! Readable)
    containerDepthBudget = maxContainerDepth;
  }
  
  public func deserialize_len() throws -> Int64 {
    let value:Int64 = reader.readInt64()
    if (value < 0 || value > Int.max) {
      throw BincodeDeserializerError.bincodeDeserializerException(issue: "Incorrect length value")
    }
    return value
  }
  
  public func deserialize_variant_index() -> Int {
    return Int(reader.readInt())
  }
  
  func deserialize_char() throws -> Character {
    throw BinaryDeserializerError.deserializationException(issue: "Not implemented: char deserialization")
  }
  
  func deserialize_f32() -> Float {
    return reader.readFloat()
  }
  
  func deserialize_f64() -> Double {
    return reader.readDouble()
  }
  
  func increase_container_depth() throws {
    if containerDepthBudget == 0 {
      throw BinaryDeserializerError.deserializationException(issue: "Exceeded maximum container depth")
    }
    containerDepthBudget -= 1;
  }
  
  func decrease_container_depth() {
    containerDepthBudget += 1
  }
  
  func deserialize_str() throws -> String {
    let len:Int64 = try deserialize_len()
    if len < 0 || len > Int.max {
      throw BinaryDeserializerError.deserializationException(issue: "Incorrect length value for Swift string")
    }
    let content: [UInt8] = reader.readBytes(count: (Int)(len))
    if content.count < len {
      throw BinaryDeserializerError.deserializationException(issue: "Need len - \(content.count) more bytes for string")
    }
    return String(bytes: content, encoding: .utf8)!
  }
  
  func deserialize_bytes() throws -> [UInt8] {
    let len:Int64 = try deserialize_len()
    if len < 0 || len > Int.max {
      throw BinaryDeserializerError.deserializationException(issue: "Incorrect length value for Swift array")
    }
    let content:[UInt8] = reader.readBytes(count: (Int)(len))
    if content.count < len {
      throw BinaryDeserializerError.deserializationException(issue: "Need  \(len) - \(content.count) more bytes for byte array")
    }
    return content
  }
  
  func deserialize_bool() throws -> Bool {
    return reader.readBool()
  }
  
  func deserialize_unit() -> Unit {
    return Unit()
  }
  
  func deserialize_u8() -> UInt8 {
    return reader.readUInt8()
  }
  
  func deserialize_u16() -> UInt16 {
    return reader.readUInt16()
  }
  
  func deserialize_u32() -> UInt32 {
    return reader.readUInt()
  }
  
  func deserialize_u64() -> UInt64 {
    return reader.readUInt64()
  }
  
  func deserialize_u128() throws -> BigInt {
    let signed: BigInt =  try deserialize_i128()
    if signed >= 0 {
      return signed
    } else {
      return signed + (BigInt(1) << 128)
    }
  }
  
  func deserialize_i8() -> Int8 {
    return reader.readInt8()
  }
  
  func deserialize_i16() -> Int16 {
    return reader.readInt16()
  }
  
  func deserialize_i32() -> Int {
    return Int(reader.readInt())
  }
  
  func deserialize_i64() ->Int64 {
    return reader.readInt64()
  }
  
  func deserialize_i128() throws -> BigInt {
    let content: [UInt8]  = reader.readBytes(count: 16)
    if content.count < 16 {
      throw BinaryDeserializerError.deserializationException(issue: "Need more bytes to deserialize 128-bit integer")
    }
    return BigInt(content.map{UInt($0)})
  }
  
  func deserialize_option_tag() throws -> Bool {
    let value: UInt8 = reader.readUInt8()
    switch value {
    case 0: return false
    case 1: return true
    default: throw BinaryDeserializerError.deserializationException(issue: "Incorrect value for Option tag: \(value)")
    }
  }
}
